django/contrib/auth/hashers.py

```diff
@@ -247,31 +262,43 @@ def encode(self, password, salt, iterations=None):
         hash = base64.b64encode(hash).decode('ascii').strip()
         return "%s$%d$%s$%s" % (self.algorithm, iterations, salt, hash)
 
-    def verify(self, password, encoded):
+    def decode(self, encoded):
         algorithm, iterations, salt, hash = encoded.split('$', 3)
         assert algorithm == self.algorithm
-        encoded_2 = self.encode(password, salt, int(iterations))
+        return {
+            'algorithm': algorithm,
+            'iterations': int(iterations),
+            'salt': salt,
+            'hash': hash,
+        }
+
+    def verify(self, password, encoded):
+        decoded_entry = self.decode(encoded)
+        assert decoded_entry['algorithm'] == self.algorithm
```

Comment:
I am currently giving the PR a full final review and I think we can drop those assertions now that they are done in decode already, what do you think? (same for the assertion in `safe_summary` and other hashers)

Reply:
Seems reasonable to me. Added a commit to remove all `assert decoded_entry['algorithm'] == self.algorithm` lines. One assertion was in `must_update` and I removed that one as well.